File: current_session/1091.py
"""
1091
"""
from typing import List
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        # BFS
        # start from 'end'
        # move to nb and +1 if see neighbor = 0
        # finish until we see 0, 0

        if grid[0][0] == 1 or grid[-1][-1] == 1:
            return -1
        
        n = len(grid)
        self.seen = [[-1 for _ in range(n)] for _ in range(n)]
        stack = [(n-1,n-1,1)]
        outer = 0
        while stack and outer < n**n:
            tmp = []            # next layer
            # stack.pop()
            for i,j,cnt in stack:
                if i==j==0:
                    return cnt
                for ni, nj in [(i+1,j+1),(i+1,j),(i+1,j-1),(i,j+1),(i,j-1),(i-1,j+1),(i-1,j),(i-1,j-1)]:
                    if 0<=ni<n and 0<=nj<n and grid[ni][nj] == 0 and self.seen[ni][nj] == -1:
                        self.seen[ni][nj] = 1
                        tmp += (ni,nj,cnt+1),
            stack = tmp
            outer += 1

        return -1
                    
print(Solution().shortestPathBinaryMatrix([[0,1,1,0,0,0],[0,1,0,1,1,0],[0,1,1,0,1,0],[0,0,0,1,1,0],[1,1,1,1,1,0],[1,1,1,1,1,0]]))
print(Solution().shortestPathBinaryMatrix([[0,1,1,1,1,1,1,1],[0,1,1,0,0,0,0,0],[0,1,0,1,1,1,1,0],[0,1,0,1,1,1,1,0],[0,1,1,0,0,1,1,0],[0,1,1,1,1,0,1,0],[0,0,0,0,0,1,1,0],[1,1,1,1,1,1,1,0]]))

File: current_session/c++/1091.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;
// use bfs
// use visited
class Solution {
public:
  int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
    int H = grid.size(), W = grid[0].size();
    queue<vector<int>> q;
    if (grid[0][0] == 1 || grid.back().back() == 1) return -1;
    q.push({0,0,1});
    while (!q.empty()) {
      vector<int> p = q.front();
      q.pop();
      int i = p[0], j=p[1], cnt=p[2];
      if (i==H-1 && j==W-1) return cnt;
      helper(grid, q, i+1, j+1, cnt, H, W);
      helper(grid, q, i+1, j, cnt, H, W);
      helper(grid, q, i+1, j-1, cnt, H, W);
      helper(grid, q, i, j+1, cnt, H, W);
      helper(grid, q, i, j-1, cnt, H, W);
      helper(grid, q, i-1, j+1, cnt, H, W);
      helper(grid, q, i-1, j, cnt, H, W);
      helper(grid, q, i-1, j-1, cnt, H, W);
    }
    return -1;
  }
  void helper(vector<vector<int>>& grid, queue<vector<int>> &nbs, int a, int b, int c, int H, int W) {
    if (0 <= a && a < H && 0 <= b && b < W && grid[a][b] == 0) {
      // cout << "a:" << a << "  b:" << b << endl;
      nbs.push(vector<int>{a,b,c+1});
      grid[a][b] = 1;
    }
  }
};